src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs

```diff
@@ -305,5 +405,77 @@ await _usersAndRolesDbContext.UserRoles
                         .OnSuccess(() => _userRoleService.AddGlobalRole(userId, roleId));
                 });
         }
+
+        private async Task<Either<ActionResult, Unit>> ValidateEmail(string email)
+        {
+            if (string.IsNullOrWhiteSpace(email))
+            {
+                return ValidationActionResult(InvalidEmailAddress);
+            }
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Unit>> ValidateUserDoesNotExist(string email)
+        {
+            if (_usersAndRolesDbContext.Users.Any(u => u.Email.ToLower() == email.ToLower()))
+            {
+                return ValidationActionResult(UserAlreadyExists);
+            }
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Publication>> ValidateReleaseIds(Publication publication,
+            List<Guid> releaseIds)
+        {
+            await _contentDbContext
+                .Entry(publication)
+                .Collection(p => p.Releases)
+                .LoadAsync();
+
+            var publicationReleaseIds = publication
+                .ListActiveReleases()
+                .Select(r => r.Id)
+                .ToList();
+
+            var intersectedReleaseIds = releaseIds.Intersect(publicationReleaseIds).ToList();
+
+            if (!intersectedReleaseIds.All(releaseIds.Contains)) // if lists don't contain same elements
+            {
+                return ValidationActionResult(NotAllReleasesBelongToPublication);
+            }
+
+            return publication;
+        }
+
+        private async Task<Either<ActionResult, Unit>> SendContributorInviteEmail(
+            string publicationTitle, List<Guid> releaseIds, string email)
+        {
+            var template = _configuration.GetValue<string>("NotifyContributorTemplateId");
+            var scheme = _httpContextAccessor.HttpContext.Request.Scheme;
+            var host = _httpContextAccessor.HttpContext.Request.Host;
+            var url = $"{scheme}://{host}/";
+
+            var releaseTitleBullets = await _contentDbContext.Releases
+                .AsAsyncEnumerable()
+                .Where(r => releaseIds.Contains(r.Id))
+                .Select(r => $"* {r.Title}")
+                .ToListAsync();
+            var releaseList = string.Join("\n", releaseTitleBullets);
+            if (releaseList.IsNullOrEmpty())
```

Comment:
Can/should this ever occur?

Reply:
It shouldn't occur. I decided to include this defensively. It could throw a validation error instead?

Reply:
It just confused me and immediately made me wonder what situation allows that but it turns out it's hypothetical.

I don't do this type of defensive coding myself because it allows problems causing scenarios like this elsewhere to go unseen.

An end user isn't necessarily going to mention to us that they are receiving emails about being granted no releases.

I can understand being defensive if this situation could arise because of old or bad data that isn't worth tidying up but this situation can only arise as a result of there being something wrong with this new bit of functionality added here allowing users to be invited with no releases which should never happen.

If something could blow up as a result of there being no Releases then I'd say it would better to do so, so that we find out about it.

I see in this particular case though that I don't think anything could blow up, the email would probably just contain a blank value.

If you want to be defensive about it and give us a better chance of spotting an issue then you could begin with something like

```
if (!releases.Any()) { throw new ArgumentException("List of releases cannot be empty", nameof(releases)); }
```

and then there should be no need for the check later on.

Reply:
Makes sense. I've changed it. In the future I'll throw Exceptions for this kind of thing.